syntax = "proto3";

package avninv.catalog;

/** Data structures **/
message PartAttributes {
    string attribute = 1;
    string unit = 2;
    optional double numeric_value = 3;
    string value = 4;
}

message PartSupplier {
    string supplier = 1;
    string supplier_part_number = 2;
    string url = 3;
}

message PartDetails {
    string id = 1;
    string manufacturer_part_number = 2;
    string type = 3;

    string description = 4;

    repeated PartAttributes attributes = 5;
    repeated PartSupplier suppliers = 6;
}

/** Requests **/
message AddPartRequest {
    PartDetails details = 1;
}

message AddPartResponse {
    string id = 1;
}

message GetPartRequest {
    string id = 1;
}

message GetPartResponse {
    PartDetails details = 1;
}

/** Service **/
service Catalog {
    rpc AddPart(AddPartRequest) returns (AddPartResponse);
    rpc GetPart(GetPartRequest) returns (GetPartResponse);
}